Source/WarpX.cpp

```diff
@@ -489,6 +503,61 @@ WarpX::PrintGlobalWarnings(const std::string& when)
     amrex::Print() << warn_string;
 }
 
+void
+WarpX::SignalSetFlag(int signal_number)
+{
+    signal_received_flags[signal_number] = true;
+}
+
+void
+WarpX::InitSignalHandling()
+{
+    bool have_checkpoint_diagnostic = false;
+
+    ParmParse pp("diagnostics");
+    std::vector<std::string> diags_names;
+    pp.queryarr("diags_names", diags_names);
+
+    for (const auto &diag : diags_names) {
+        ParmParse dd(diag);
+        std::string format;
+        dd.query("format", format);
+        if (format == "checkpoint") {
+            have_checkpoint_diagnostic = true;
+            break;
+        }
+    }
+
+#if defined(__linux__) || defined(__APPLE__)
+    struct sigaction sa;
+    sigemptyset(&sa.sa_mask);
+    for (int i = 0; i < NUM_SIGNALS; ++i) {
+        signal_received_flags[i] = false;
+        if (signal_conf_requests_checkpoint[i] || signal_conf_requests_break[i]) {
+            if (ParallelDescriptor::MyProc() == 0) {
+                sa.sa_handler = &WarpX::SignalSetFlag;
+            } else {
+                sa.sa_handler = SIG_IGN;
+            }
+            int result = sigaction(i, &sa, nullptr);
+            if (result != 0) {
+                Abort("Failed to install signal handler for a configured signal");
```

Comment:
Here and below, we can make error messages more uniform via `WARPX_ALWAYS_ABORT_WITH_MESSAGE` #2952